File: app/src/main/java/com/settlex/android/util/network/NetworkMonitor.java
package com.settlex.android.util.network;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.settlex.android.SettleXApp;

/**
 * Monitors network connectivity status and provides observable LiveData for network availability.
 * Call startNetworkCallback() once during application initialization to begin monitoring.
 */
public class NetworkMonitor {
    private static final MutableLiveData<Boolean> networkStatus = new MutableLiveData<>();

    private NetworkMonitor() {
        // Prevent instantiation
    }

    /**
     * Registers network callback to monitor connectivity changes
     * and provides initial status.
     */
    public static void startNetworkCallback() {
        Context context = SettleXApp.getAppContext();
        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if (cm == null) return;

        ConnectivityManager.NetworkCallback callback = new ConnectivityManager.NetworkCallback() {
            @Override
            public void onAvailable(@NonNull Network network) {
                networkStatus.postValue(true);
            }

            @Override
            public void onLost(@NonNull Network network) {
                networkStatus.postValue(false);
            }

            @Override
            public void onUnavailable() {
                networkStatus.postValue(false);
            }
        };

        cm.registerDefaultNetworkCallback(callback);

        // Set initial network state
        NetworkCapabilities caps = cm.getNetworkCapabilities(cm.getActiveNetwork());
        boolean isConnected = caps != null && caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
        networkStatus.setValue(isConnected);
    }

    /**
     * Returns LiveData observable for network connectivity status.
     * @return LiveData<Boolean> where true indicates internet connectivity is available
     */
    public static LiveData<Boolean> getNetworkStatus() {
        return networkStatus;
    }
}

File: app/src/main/java/com/settlex/android/util/network/NetworkMonitor.kt
package com.settlex.android.util.network

import android.content.Context
import android.net.ConnectivityManager
import android.net.ConnectivityManager.NetworkCallback
import android.net.Network
import android.net.NetworkCapabilities
import com.settlex.android.SettleXApp
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

object NetworkMonitor {
    private val _networkStatus: MutableStateFlow<Boolean> = MutableStateFlow(false)
    val networkStatus: StateFlow<Boolean> = _networkStatus.asStateFlow()

    fun startNetworkCallback() {
        val context = SettleXApp.appContext
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager?
        if (cm == null) return

        val callback: NetworkCallback = object : NetworkCallback() {
            override fun onAvailable(network: Network) {
                _networkStatus.tryEmit(true)
            }

            override fun onLost(network: Network) {
                _networkStatus.tryEmit(false)
            }

            override fun onUnavailable() {
                _networkStatus.tryEmit(false)
            }
        }

        cm.registerDefaultNetworkCallback(callback)

        // Set initial network state
        val activeNetwork = cm.activeNetwork
        val caps = cm.getNetworkCapabilities(activeNetwork)
        _networkStatus.value =
            caps?.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) == true
    }
}